source/distributed/shared_tria.cc

```diff
@@ -76,14 +107,85 @@ namespace parallel
                 cell->set_subdomain_id(numbers::artificial_subdomain_id);
             }
+
+          // loop over all cells in multigrid hierarchy and mark artificial:
+          if (settings & construct_multigrid_hierarchy)
+            {
+              std_cxx11::function<bool (const typename parallel::shared::Triangulation<dim,spacedim>::cell_iterator &)> predicate
+                = IteratorFilters::LocallyOwnedLevelCell();
+              for (unsigned int lvl=0; lvl<this->n_levels(); ++lvl)
+                {
+                  const std::vector<typename parallel::shared::Triangulation<dim,spacedim>::cell_iterator>
+                  level_halo_layer_vector = GridTools::compute_cell_halo_layer_on_level (*this, predicate, lvl);
+                  std::set<typename parallel::shared::Triangulation<dim,spacedim>::cell_iterator>
+                  level_halo_layer(level_halo_layer_vector.begin(), level_halo_layer_vector.end());
+
+                  typename parallel::shared::Triangulation<dim,spacedim>::cell_iterator
+                  cell = this->begin(lvl),
+                  endc = this->end(lvl);
+                  for (; cell != endc; cell++)
+                    {
+                      // for active cells we must keep level subdomain id of all neighbors
+                      // to our subdomain, not just cells that share a vertex on the same level.
+                      // if the cells subdomain id was not set to artitficial above, we will
+                      // also keep its level subdomain id since it is either owned by this processor
+                      // or in the ghost layer of the active mesh.
+                      if (!cell->has_children() && cell->subdomain_id() != numbers::artificial_subdomain_id)
+                        continue;
+                      if (!cell->is_locally_owned_on_level() &&
+                          level_halo_layer.find(cell) == level_halo_layer.end())
+                        cell->set_level_subdomain_id(numbers::artificial_subdomain_id);
+                    }
+                }
+            }
         }
       else
         {
           // just store true subdomain ids
           for (unsigned int index=0; cell != endc; cell++, index++)
             true_subdomain_ids_of_cells[index] = cell->subdomain_id();
+        }
 
+#ifdef DEBUG
+      {
+        // Assert that each cell is owned by a processor
+        unsigned int n_my_cells = 0;
+        typename parallel::shared::Triangulation<dim,spacedim>::active_cell_iterator
+        cell = this->begin_active(),
+        endc = this->end();
+        for (; cell!=endc; ++cell)
+          if (cell->is_locally_owned())
+            n_my_cells += 1;
+
+        unsigned int total_cells;
+        int ierr = MPI_Allreduce (&n_my_cells, &total_cells, 1,
+                                  MPI_UNSIGNED, MPI_SUM, this->get_communicator());
+        AssertThrowMPI(ierr);
+
+        Assert(total_cells == this->n_active_cells(),
```

Comment:
i hit this Assert, i think it's wrong as `total_cells` is sum of all locally owned, whereas `n_active_cells()` is a MPI-local value. 

But there is another issue: `n_global_active_cells` and `n_active_cells` uses `number_cache`, but this case is update in `execute_coarsening_and_refinement()` **after** `partition()`. I see if I can fix this...